pkg/quarterdeck/auth.go

```diff
@@ -254,7 +254,60 @@ func (s *Server) Authenticate(c *gin.Context) {
 	c.JSON(http.StatusOK, out)
 }
 
-// TODO: implement the refresh endpoint
+// Refresh is oriented to machine users that have a refresh token and need to obtain
+// a new access token and refresh token to continue using ensign systems
 func (s *Server) Refresh(c *gin.Context) {
-	c.JSON(http.StatusNotImplemented, api.ErrorResponse("not yet implemented"))
+	var (
+		err error
+		in  *api.RefreshRequest
+		out *api.LoginReply
+	)
+
+	if err = c.BindJSON(&in); err != nil {
+		log.Warn().Err(err).Msg("could not parse refresh request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not parse request"))
+		return
+	}
+
+	// Check to see if the refresh token is included in the request
+	if in.RefreshToken == "" {
+		log.Debug().Msg("missing refresh token from request request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("missing credentials"))
+		return
+	}
+
+	// verify the refresh token
+	claims, err := s.tokens.Verify(in.RefreshToken)
+	if err != nil {
+		log.Error().Err(err).Msg("could not verify refresh token")
+		c.JSON(http.StatusUnauthorized, api.ErrorResponse("could not verify refresh token"))
+		return
+	}
+
+	// get the user from the database using the ID
+	user, err := models.GetUser(c, claims.Subject)
+	if err != nil {
+		log.Error().Err(err).Msg("could not retrieve user from claims")
+		c.JSON(http.StatusUnauthorized, api.ErrorResponse("could not retrieve user from claims"))
+		return
+	}
+
```

Comment:
I'm glad we're getting the user from the database, that's a good check to make sure that the user still exists or hasn't been deleted (we do not want to keep authenticating a deleted user). I'm thinking that we should also fetch the claims from the database again rather than use the claims from the refresh token; this will ensure that if the user's permissions (or name or email) have changed between login and refresh that those new claims are what gets sent rather than the old claims. 

Are there any other checks or security related things you can think of that we should do in this PR?